src/library/scala/collection/mutable/TreeMap.scala

```diff
@@ -132,13 +154,15 @@ sealed class TreeMap[K, V] private (tree: RB.Tree[K, V])(implicit val ordering:
 
     override def get(key: K) = if (isInsideViewBounds(key)) RB.get(tree, key) else None
 
-    override def iterator = RB.iterator(tree, from, until)
-    override def keysIteratorFrom(start: K) = RB.keysIterator(tree, pickLowerBound(Some(start)), until)
-    override def iteratorFrom(start: K) = RB.iterator(tree, pickLowerBound(Some(start)), until)
-    override def valuesIteratorFrom(start: K) = RB.valuesIterator(tree, pickLowerBound(Some(start)), until)
-
-    override def size = iterator.length
-    override def isEmpty = !iterator.hasNext
+    override def iterator = if (RB.size(tree) == 0) Iterator.empty else RB.iterator(tree, from, until)
+    override def keysIterator: Iterator[K] = if (RB.size(tree) == 0) Iterator.empty else RB.keysIterator(tree, from, until)
+    override def valuesIterator: Iterator[V] = if (RB.size(tree) == 0) Iterator.empty else RB.valuesIterator(tree, from, until)
+    override def keysIteratorFrom(start: K) = if (RB.size(tree) == 0) Iterator.empty else RB.keysIterator(tree, pickLowerBound(Some(start)), until)
+    override def iteratorFrom(start: K) = if (RB.size(tree) == 0) Iterator.empty else RB.iterator(tree, pickLowerBound(Some(start)), until)
+    override def valuesIteratorFrom(start: K) = if (RB.size(tree) == 0) Iterator.empty else RB.valuesIterator(tree, pickLowerBound(Some(start)), until)
+    override def size = if (RB.size(tree) == 0) 0 else iterator.length
+    override def knownSize: Int = if (RB.size(tree) == 0) 0 else -1
```

Comment:
`RB.size` is not a constant-time operation. Is it fast enough to be called in `knownSize` (and to a lesser extent in the other methods here)?

Reply:
There are two overloads of `RB.size`. I'm using the one that exposes the cached size of the tree.

```
def size(tree: Tree[_, _]): Int = tree.size
final class Tree[A, B](var root: Node[A, B], var size: Int)
```

The other variant is only used in tests via `RB.isValid`.